src/libxmljs.cc

```diff
@@ -97,8 +97,26 @@ char* xmlMemoryStrdupWrap(const char* str)
     return res;
 }
 
+// callback function for `xmlDeregisterNodeDefault`
+void xmlDeregisterNodeCallback(xmlNode* xml_obj)
+{
+    if (xml_obj->_private)
+    {
+        XmlNode* node = static_cast<XmlNode*>(xml_obj->_private);
+
+        // flag the XmlNode object as freed
+        node->freed = true;
+
+        // save a reference to the doc so we can still `unref` it
+        node->doc = xml_obj->doc;
+    }
+    return;
+}
+
 LibXMLJS::LibXMLJS()
 {
+    // set the callback for when a node is about to be freed
+    xmlDeregisterNodeDefault(xmlDeregisterNodeCallback);
```

Comment:
What about [xmlThrDefDeregisterNodeDefault](http://www.xmlsoft.org/html/libxml-globals.html#xmlThrDefDeregisterNodeDefault) ? Are we guaranteed to always work in the same thread ?